File: frontend/src/api/creditPortfolio.ts
import apiClient from './client'

// Types will be defined later in types/creditPortfolio.ts
export interface FinOrganization {
  id: number
  name: string
  inn?: string
  is_active: boolean
  department_id: number
  created_at: string
  updated_at?: string
}

export interface FinBankAccount {
  id: number
  account_number: string
  bank_name?: string
  is_active: boolean
  department_id: number
  created_at: string
  updated_at?: string
}

export interface FinContract {
  id: number
  contract_number: string
  contract_date?: string
  contract_type?: string
  counterparty?: string
  is_active: boolean
  department_id: number
  created_at: string
  updated_at?: string
}

export interface FinReceipt {
  id: number
  operation_id: string
  organization_id: number
  bank_account_id?: number
  contract_id?: number
  operation_type?: string
  accounting_account?: string
  document_number?: string
  document_date?: string
  payer?: string
  payer_account?: string
  settlement_account?: string
  contract_date?: string
  currency: string
  amount: number
  commission?: number
  payment_purpose?: string
  responsible_person?: string
  comment?: string
  is_active: boolean
  department_id: number
  created_at: string
  updated_at?: string
}

export interface FinExpense {
  id: number
  operation_id: string
  organization_id: number
  bank_account_id?: number
  contract_id?: number
  operation_type?: string
  accounting_account?: string
  document_number?: string
  document_date?: string
  recipient?: string
  recipient_account?: string
  debit_account?: string
  contract_date?: string
  currency: string
  amount: number
  expense_article?: string
  payment_purpose?: string
  responsible_person?: string
  comment?: string
  tax_period?: string
  unconfirmed_by_bank: boolean
  is_active: boolean
  department_id: number
  created_at: string
  updated_at?: string
}

export interface CreditPortfolioSummary {
  total_receipts: number
  total_expenses: number
  net_balance: number
  active_contracts_count: number
  total_interest: number
  total_principal: number
}

export interface MonthlyStats {
  month: string
  receipts: number
  expenses: number
  net: number
}

export const creditPortfolioApi = {
  // ==================== Organizations ====================
  getOrganizations: async (params?: {
    skip?: number
    limit?: number
    department_id?: number
    is_active?: boolean
  }) => {
    const { data } = await apiClient.get('/credit-portfolio/organizations', { params })
    return data
  },

  getOrganization: async (id: number) => {
    const { data } = await apiClient.get(`/credit-portfolio/organizations/${id}`)
    return data
  },

  // ==================== Bank Accounts ====================
  getBankAccounts: async (params?: {
    skip?: number
    limit?: number
    department_id?: number
    is_active?: boolean
  }) => {
    const { data } = await apiClient.get('/credit-portfolio/bank-accounts', { params })
    return data
  },

  getBankAccount: async (id: number) => {
    const { data } = await apiClient.get(`/credit-portfolio/bank-accounts/${id}`)
    return data
  },

  // ==================== Contracts ====================
  getContracts: async (params?: {
    skip?: number
    limit?: number
    department_id?: number
    is_active?: boolean
    contract_type?: string
  }) => {
    const { data } = await apiClient.get('/credit-portfolio/contracts', { params })
    return data
  },

  getContract: async (id: number) => {
    const { data } = await apiClient.get(`/credit-portfolio/contracts/${id}`)
    return data
  },

  // ==================== Receipts ====================
  getReceipts: async (params?: {
    skip?: number
    limit?: number
    department_id?: number
    organization_id?: number
    bank_account_id?: number
    contract_id?: number
    date_from?: string
    date_to?: string
  }): Promise<FinReceipt[]> => {
    const { data } = await apiClient.get('/credit-portfolio/receipts', { params })
    return data
  },

  getReceipt: async (id: number): Promise<FinReceipt> => {
    const { data } = await apiClient.get(`/credit-portfolio/receipts/${id}`)
    return data
  },

  // ==================== Expenses ====================
  getExpenses: async (params?: {
    skip?: number
    limit?: number
    department_id?: number
    organization_id?: number
    bank_account_id?: number
    contract_id?: number
    date_from?: string
    date_to?: string
  }): Promise<FinExpense[]> => {
    const { data } = await apiClient.get('/credit-portfolio/expenses', { params })
    return data
  },

  getExpense: async (id: number): Promise<FinExpense> => {
    const { data } = await apiClient.get(`/credit-portfolio/expenses/${id}`)
    return data
  },

  // ==================== Analytics ====================
  getSummary: async (params?: {
    department_id?: number
    date_from?: string
    date_to?: string
  }): Promise<CreditPortfolioSummary> => {
    const { data } = await apiClient.get('/credit-portfolio/summary', { params })
    return data
  },

  getMonthlyStats: async (params?: {
    department_id?: number
    year?: number
  }): Promise<MonthlyStats[]> => {
    const { data } = await apiClient.get('/credit-portfolio/monthly-stats', { params })
    return data
  },

  getContractStats: async (params?: {
    department_id?: number
    date_from?: string
    date_to?: string
  }) => {
    const { data } = await apiClient.get('/credit-portfolio/contract-stats', { params })
    return data
  },

  getOrganizationStats: async (params?: {
    department_id?: number
    date_from?: string
    date_to?: string
  }) => {
    const { data } = await apiClient.get('/credit-portfolio/organization-stats', { params })
    return data
  },

  // ==================== Import ====================
  triggerImport: async () => {
    const { data } = await apiClient.post('/credit-portfolio/import/trigger')
    return data
  },

  getImportLogs: async (params?: {
    skip?: number
    limit?: number
    department_id?: number
  }) => {
    const { data } = await apiClient.get('/credit-portfolio/import/logs', { params })
    return data
  },
}

File: backend/app/schemas/credit_portfolio.py
"""
Pydantic schemas for Credit Portfolio
"""
from datetime import date, datetime
from decimal import Decimal
from typing import Optional
from pydantic import BaseModel, Field


# ==================== FinOrganization ====================

class FinOrganizationBase(BaseModel):
    name: str = Field(..., description="Название организации")
    inn: Optional[str] = Field(None, description="ИНН")
    is_active: bool = Field(True, description="Активность")


class FinOrganizationCreate(FinOrganizationBase):
    department_id: Optional[int] = Field(None, description="ID отдела (опционально для ADMIN/MANAGER)")


class FinOrganizationUpdate(BaseModel):
    name: Optional[str] = None
    inn: Optional[str] = None
    is_active: Optional[bool] = None


class FinOrganizationInDB(FinOrganizationBase):
    id: int
    department_id: int
    created_at: datetime
    updated_at: Optional[datetime]

    class Config:
        from_attributes = True


# ==================== FinBankAccount ====================

class FinBankAccountBase(BaseModel):
    account_number: str = Field(..., description="Номер счета")
    bank_name: Optional[str] = Field(None, description="Название банка")
    is_active: bool = Field(True, description="Активность")


class FinBankAccountCreate(FinBankAccountBase):
    department_id: Optional[int] = Field(None)


class FinBankAccountUpdate(BaseModel):
    account_number: Optional[str] = None
    bank_name: Optional[str] = None
    is_active: Optional[bool] = None


class FinBankAccountInDB(FinBankAccountBase):
    id: int
    department_id: int
    created_at: datetime
    updated_at: Optional[datetime]

    class Config:
        from_attributes = True


# ==================== FinContract ====================

class FinContractBase(BaseModel):
    contract_number: str = Field(..., description="Номер договора")
    contract_date: Optional[date] = Field(None, description="Дата договора")
    contract_type: Optional[str] = Field(None, description="Тип договора (Кредит, Заем)")
    counterparty: Optional[str] = Field(None, description="Контрагент")
    is_active: bool = Field(True, description="Активность")


class FinContractCreate(FinContractBase):
    department_id: Optional[int] = None


class FinContractUpdate(BaseModel):
    contract_number: Optional[str] = None
    contract_date: Optional[date] = None
    contract_type: Optional[str] = None
    counterparty: Optional[str] = None
    is_active: Optional[bool] = None


class FinContractInDB(FinContractBase):
    id: int
    department_id: int
    created_at: datetime
    updated_at: Optional[datetime]

    class Config:
        from_attributes = True


# ==================== FinReceipt ====================

class FinReceiptBase(BaseModel):
    operation_id: str
    organization_id: int
    bank_account_id: Optional[int] = None
    contract_id: Optional[int] = None
    operation_type: Optional[str] = None
    accounting_account: Optional[str] = None
    document_number: Optional[str] = None
    document_date: Optional[date] = None
    payer: Optional[str] = None
    payer_account: Optional[str] = None
    settlement_account: Optional[str] = None
    contract_date: Optional[date] = None
    currency: str = "RUB"
    amount: Decimal
    commission: Optional[Decimal] = None
    payment_purpose: Optional[str] = None
    responsible_person: Optional[str] = None
    comment: Optional[str] = None


class FinReceiptCreate(FinReceiptBase):
    department_id: Optional[int] = None


class FinReceiptInDB(FinReceiptBase):
    id: int
    department_id: int
    is_active: bool
    created_at: datetime
    updated_at: Optional[datetime]

    class Config:
        from_attributes = True


# ==================== FinExpense ====================

class FinExpenseBase(BaseModel):
    operation_id: str
    organization_id: int
    bank_account_id: Optional[int] = None
    contract_id: Optional[int] = None
    operation_type: Optional[str] = None
    accounting_account: Optional[str] = None
    document_number: Optional[str] = None
    document_date: Optional[date] = None
    recipient: Optional[str] = None
    recipient_account: Optional[str] = None
    debit_account: Optional[str] = None
    contract_date: Optional[date] = None
    currency: str = "RUB"
    amount: Decimal
    expense_article: Optional[str] = None
    payment_purpose: Optional[str] = None
    responsible_person: Optional[str] = None
    comment: Optional[str] = None
    tax_period: Optional[str] = None
    unconfirmed_by_bank: bool = False


class FinExpenseCreate(FinExpenseBase):
    department_id: Optional[int] = None


class FinExpenseInDB(FinExpenseBase):
    id: int
    department_id: int
    is_active: bool
    created_at: datetime
    updated_at: Optional[datetime]

    class Config:
        from_attributes = True


# ==================== FinExpenseDetail ====================

class FinExpenseDetailBase(BaseModel):
    expense_operation_id: str
    contract_number: Optional[str] = None
    repayment_type: Optional[str] = None
    settlement_account: Optional[str] = None
    advance_account: Optional[str] = None
    payment_type: Optional[str] = None
    payment_amount: Optional[Decimal] = None
    settlement_rate: Decimal = Decimal("1.0")
    settlement_amount: Optional[Decimal] = None
    vat_amount: Optional[Decimal] = None
    expense_amount: Optional[Decimal] = None
    vat_in_expense: Optional[Decimal] = None


class FinExpenseDetailCreate(FinExpenseDetailBase):
    department_id: Optional[int] = None


class FinExpenseDetailInDB(FinExpenseDetailBase):
    id: int
    department_id: int
    created_at: datetime

    class Config:
        from_attributes = True


# ==================== FinImportLog ====================

class FinImportLogInDB(BaseModel):
    id: int
    import_date: datetime
    source_file: Optional[str]
    table_name: Optional[str]
    rows_inserted: int = 0
    rows_updated: int = 0
    rows_failed: int = 0
    status: Optional[str]
    error_message: Optional[str]
    processed_by: Optional[str]
    processing_time_seconds: Optional[Decimal]
    department_id: int

    class Config:
        from_attributes = True


# ==================== Summary & Stats ====================

class CreditPortfolioSummary(BaseModel):
    """Summary statistics for credit portfolio"""
    total_receipts: Decimal = Field(..., description="Всего поступлений")
    total_expenses: Decimal = Field(..., description="Всего списаний")
    net_balance: Decimal = Field(..., description="Чистый баланс")
    active_contracts_count: int = Field(..., description="Активных договоров")
    total_interest: Decimal = Field(..., description="Всего процентов")
    total_principal: Decimal = Field(..., description="Всего тела кредита")


class MonthlyStats(BaseModel):
    """Monthly statistics"""
    month: str
    receipts: Decimal
    expenses: Decimal
    net: Decimal
